### PR TITLE
meta: backup

### PR DESCRIPTION
**Pull request by jiefenghuang**
## Goal
A new binary backup method enables more efficient and memory-saving dump and load operations, producing smaller backups. It is designed for large-scale file system backups and migrations.

## Limited
- Dump: For DB (single concurrency) and TiKV, data consistency is ensured through snapshot reads. In other scenarios, such as DB (with multiple concurrency) and Redis, backups need to be exported in read-only mode.

## Performance
data set: 1m files
goos: linux
goarch: amd64
cpu: AMD Ryzen 7 7840HS w/ Radeon 780M Graphics
- mysql
  - dumpv1(non-fast): 49.87s
  - dumpv1(fast): 18.61s
  - dumpv2(CoNum=1): 16.93s
  - dumpv2(CoNum=10): 4.18s
  - loadv1: 72.06s
  - loadv2(CoNum=10): 32.70s
- redis 
  - dumpv1: 4.52s, 3.96s ...
  - dumpv2(CoNum=1): 3.65s
  - dumpv2(CoNum=10): 3.45s
  - loadv1: 7.30s
  - loadv2(CoNum=10): 1.32s
- tikv (tiup playground)
  - dumpv1(non-fast): 54.60s
  - dumpv1(fast): 7.16s
  - dumpv2(CoNum=1):  1.81s
  - dumpv2(CoNum>1):  TODO: need a TiKV cluster with multiple disks to check this
  - loadv1: 9.64s
  - loadv2(CoNum=10): 5.42s
```shell
BenchmarkLoadDumpV2/Load_mysql-16                      1        72060549327 ns/op       9986896728 B/op 137978053 allocs/op
BenchmarkLoadDumpV2/Dump_non-fast_mode_mysql-16                        1        49878605404 ns/op       19085003992 B/op     554803266 allocs/op
BenchmarkLoadDumpV2/Dump_fast_mode_mysql-16                            1        18618667068 ns/op       7130212416 B/op      292465453 allocs/op
BenchmarkLoadDumpV2/DumpV2_CoNum=1_mysql-16                            1        16930252356 ns/op       7911537616 B/op      279367159 allocs/op
BenchmarkLoadDumpV2/DumpV2_CoNum=10_mysql-16                           1        4186731500 ns/op        7781710912 B/op      278340145 allocs/op
BenchmarkLoadDumpV2/LoadV2_mysql-16                                    1        32705284621 ns/op       8056306592 B/op 169484502 allocs/op

BenchmarkLoadDumpV2/Load_redis-16                      1        7307923118 ns/op        2062659712 B/op 27617610 allocs/op
BenchmarkLoadDumpV2/Dump_non-fast_mode_redis-16                        1        4528533637 ns/op        2303354744 B/op 41546163 allocs/op
BenchmarkLoadDumpV2/Dump_fast_mode_redis-16                            1        3967532937 ns/op        2299847272 B/op 41513331 allocs/op
BenchmarkLoadDumpV2/DumpV2_CoNum=1_redis-16                            1        3659069323 ns/op        712397208 B/op   9024910 allocs/op
BenchmarkLoadDumpV2/DumpV2_CoNum=10_redis-16                           1        3456428554 ns/op        713347296 B/op   9033111 allocs/op
BenchmarkLoadDumpV2/LoadV2_redis-16                                    1        1323064794 ns/op        1063459224 B/op 15249467 allocs/op

BenchmarkLoadDumpV2/Load_tikv-16                       1        9644265753 ns/op        2708921496 B/op 28191080 allocs/op
BenchmarkLoadDumpV2/Dump_non-fast_mode_tikv-16                 1        54607733112 ns/op       14235384304 B/op        256777839 allocs/op
BenchmarkLoadDumpV2/Dump_fast_mode_tikv-16                     1        7169758572 ns/op        2514788216 B/op 34111266 allocs/op
BenchmarkLoadDumpV2/DumpV2_CoNum=1_tikv-16                     1        1812321521 ns/op        846469264 B/op   8966257 allocs/op
BenchmarkLoadDumpV2/LoadV2_tikv-16                             1        5420456001 ns/op        991102312 B/op  10047605 allocs/op
```

## Memory

- mysql (TODO: optimise)
image

- redis
image

- tikv
image

## TODO

- merge redis scan operations(node, chunks, ...)
- optimize mysql memory
- break down large transactions in TiKV into multiple smaller transactions, each with an independent but identically versioned snapshot.

### CODE REVIEW

File: pkg/meta/base.go [hunk 1]
```diff
@@ -3067,3 +3074,133 @@ func (h *dirHandler) Close() {
 	h.readOff = 0
 	h.Unlock()
 }
+
+func (m *baseMeta) DumpMetaV2(ctx Context, w io.Writer, opt *DumpOption) error {
+	opt = opt.check()
+
+	bak := NewBakFormat()
+	ch := make(chan *dumpedResult, 100)
+	wg := &sync.WaitGroup{}
```
Comment (SandyXSD) [inline pkg/meta/base.go:3083]: Not necessary? `ctx` and `ch` are enough to do the synchronization.

File: pkg/meta/base.go [hunk 2]
```diff
@@ -3067,3 +3074,133 @@ func (h *dirHandler) Close() {
...
+		})
+		if err != nil {
+			ctx.Cancel()
+		} else {
+			close(ch)
+		}
+	}()
+
+	finished := false
+	for !finished {
```
Comment (SandyXSD) [inline pkg/meta/base.go:3116]:
```
	var res *dumpedResult
	for {
		select {
		case <-ctx.Done():
			return ctx.Err()
		case res = <-ch:
		}
		if res == nil {
			break
		}
		...
	}
```

File: pkg/meta/tkv_bak.go [hunk 3]
```diff
@@ -0,0 +1,850 @@
...
+
+	logger.Debugf("dump %s num: %s", s, printSums(sums))
+	return err
+}
+
+func (s *kvMixDS) release(msg proto.Message) {
+	switch list := msg.(type) {
+	case *pb.NodeList:
+		for _, node := range list.List {
+			s.pools[0].Put(node)
```
Comment (SandyXSD) [inline pkg/meta/tkv_bak.go:507]: reset fields of node before putting back.

File: pkg/meta/redis_bak.go [hunk 4]
```diff
@@ -0,0 +1,1206 @@
...
+	dumpedBatchSeg
+}
+
+func (s *redisNodeDBS) dump(ctx Context, ch chan *dumpedResult) error {
+	meta := s.meta.(*redisMeta)
+	eg, egCtx := errgroup.WithContext(ctx)
+	eg.SetLimit(s.opt.CoNum)
+	var sum int64
+
+	if err := meta.scan(egCtx, "i*", func(iKeys []string) error {
```
Comment (SandyXSD) [inline pkg/meta/redis_bak.go:545]: Try not use the pattern scan.

File: pkg/meta/sql_bak.go [hunk 5]
```diff
@@ -0,0 +1,1050 @@
...
+
+	taskFinished := false
+	psrs := &pb.SliceRefList{List: make([]*pb.SliceRef, 0, 1024)}
+	for start := 0; !taskFinished; start += sqlBatchSize {
+		nStart := start
+		eg.Go(func() error {
+			var rows []sliceRef
+			if err := s.meta.(*dbMeta).execStmt(ctx, s.txn, func(s *xorm.Session) error {
+				rows = rows[:0]
+				return s.Where("refs != 1").Limit(sqlBatchSize, nStart).Find(&rows) // skip default refs
```
Comment (SandyXSD) [inline pkg/meta/sql_bak.go:275]: `refs > 1`?

Comment (jiefenghuang) [reply pkg/meta/sql_bak.go:275]: ref == 0 means that the part to be reclaimed is also needed.

File: pkg/meta/backup.go [hunk 6]
```diff
@@ -0,0 +1,559 @@
...
+		MinClientVersion: f.MinClientVersion,
+		MaxClientVersion: f.MaxClientVersion,
+		DirStats:         f.DirStats,
+		EnableAcl:        f.EnableACL,
+	}
+
+	if !keepSecret {
+		removeKey := func(key *string, name string) {
+			if *key == "" {
+				*key = "remove"
```
[the remark this answers is not in the document]
Comment (jiefenghuang) [reply pkg/meta/backup.go:377]: changed.

File: pkg/meta/base.go (hunk 2 above)
Comment (jiefenghuang) [reply pkg/meta/base.go:3116]: done

File: pkg/meta/tkv_bak.go (hunk 3 above)
Comment (jiefenghuang) [reply pkg/meta/tkv_bak.go:507]: done

**Pull request closed without merging**